### DISCUSSION TITLE
[Question] : social networks in-app browser

### DISCUSSION THREAD

**Discussion by ryancrunchi** [google/model-viewer#1915] [Q&A]
I tested my URLs to AR models by sharing them on social networks (fb, twitter and linkedin for now).
When you open an AR model from their native iOS app, they open it in their integrated browser. Here are the result of testing `modelViewer.canActivateAR` in different iOS browsers :
- Safari : true
- Chrome : true
- Facebook in-app : false
- Linkedin in-app : false
- Twitter in-app : true

For fb and linkedin, is this because they use a custom browser with not all features available (camera, XR etc..) ?
For twitter, is this because they use a default iOS safari view controller which enables all features ?
Can I force my links (https://) to open in safari when clicked in fb or linkedin app ?

**Comment by elalish** [comment]
I believe you are correct; we have similar problems with XR not being available in an Android WebView, which is basically equivalent. I have no idea what control you may have over where your links open.

**Comment by NeoInsignia** [comment]
Hi Ryan and Emmet.
I came here to find a solution to the same issue. It's important for me to link a working URL on in-app browsing to launch a social media campaign. What is the best way to approach this? a line of code? a link to launch a browser outside the app? any suggestions?

**Comment by felipe-chamas** [comment]
I believe it's a problem of those apps trying to open USDZ files. A simple test:

```html
<!DOCTYPE html>
<html lang="en">
  <head> </head>
  <body>
    <a
      href="https://developer.apple.com/augmented-reality/quick-look/models/drummertoy/toy_drummer.usdz"
      rel="ar"
    >
      <img
        src="https://developer.apple.com/augmented-reality/quick-look/models/drummertoy/drummertoy_2x.jpg"
        alt=""
      />
    </a>
  </body>
</html>
```
- `npx live-server`
- [localtunnel](https://github.com/localtunnel/localtunnel)
- Send generated link as message on one of these
- Click on image poster:

<img src='https://user-images.githubusercontent.com/49691924/106622687-9d8e3b80-6552-11eb-90e9-9feba0581389.jpg' width='40%' height='40%'/>

I've tried launching the link on Safari with no success. For a quick-fix I added a modal that asks the user to open in safari with image instructions, based on userAgent:
`/instagram|FBAN|FBAV|LinkedInApp/i.test(navigator.userAgent)`
It's not a good option, so following the thread for more updates, but I believe it's a problem with USDZ/Apple/in-app-browsers...

**Reply by NeoInsignia** [reply]
Thank you Felipe. Having a button re directing to system browser is a better option than displaying the 3D view only, IMO. I'll keep an eye on this in case there's a code method that allows launch the AR content directly.

**Comment by uprealdigital** [comment]
Hi! 

What about Instagram and WKWebView?

The company 8th wall states that WebAR now works with your favorite apps like Instagram which uses WKWebView with the new iOS 14.3+.
https://www.8thwall.com/blog/post/41168830873/8th-wall-now-works-across-all-major-ios-apps-including-instagram-snapchat-and-more

How is it possible that it works for 8th wall but not with the model viewer?

**Reply by elalish** [reply]
Please don't cross-post. It's much easier to answer you question in one place.

**Reply by uprealdigital** [reply]
Sorry about that @elalish !

**Comment by rawwerks** [comment]
anyone have a new/good workaround for this?  i just discovered this the hard way after switching my IG bio link to https://ar.ink, only to find that the AR cannot be launched from the built-in IG browser. 

(i am hoping that @FelipeR2U 's workaround it out of date by now, a year later.)

**Reply by elalish** [reply]
I think you'll need to file a bug on WebKit if you want their WKWebViews to work with AR. Android had a similar problem: Android WebView doesn't support WebXR, but the new WebLayer does. These things take time...

**Comment by milesgreen** [comment]
There's a few different strands to this one....

First issue; on iOS, social media app developers should really be using the `SFSafariViewController` rather than `WKWebView` for displaying third-party content. AR QuickLook works in `SFSafariViewController`. Not only that, `SFSafariViewController` is a more secure, feature-complete browsing experience. Twitter uses `SFSafariViewController`. Facebook, Instagram, LinkedIn all use `WKWebView`. There's a time and place for `WKWebView`, but viewing third-party websites is not it. Canvasing iOS app developers to switch over to `SFSafariViewController` is one option. 

Second issue; being able to break out of a WebView is another challenge. It's possible for websites to detect that they are running inside a WebView. But currently there's no easy way to break out of it. There's a feature request for WebKit to support a URL scheme for opening links in the user's default web browser, which would provide a way for websites to break out of WebViews. By all means, pile on and +1 it here: 

https://bugs.webkit.org/show_bug.cgi?id=240025

**Reply by milesgreen** [reply]
@grorg - might this be of interest?

**Reply by rawwerks** [reply]
_Canvasing iOS app developers to switch over to SFSafariViewController is one option._

@milesgreen -
Any tips on how to do this for the top social apps? This has been driving me nuts, there is no smooth way to promote an AR app/website on IG, LI, FB, TikTok. Between all of the logic for checking the User Agent and then having to explain to the user why they have to copy/paste the link or manually click the [...] to Open In Safari, easily 90% of interested traffic is just not going to make the leap from social app link to a working AR experience.

**Reply by milesgreen** [reply]
> Any tips on how to do this

That's a great question @rawwerks. 

For what it's worth, I've been submitting feedback and bug forms for the main platforms. But I imagine those suggestions go into a blackhole somewhere. 

Seems like most of the social apps don't have public issue trackers / forums. A lot of the browsers do though, so that's where I've been logging a lot of browser-related issues. 

I've been trying to make connections with people inside those organisations to highlight the issue. Slow progress there. 

Per the comment below, there's an issue logged with WebKit to bring compatibility to WebView - which is what most of the social apps are using: https://bugs.webkit.org/show_bug.cgi?id=240117

@grorg is in the loop and fighting the good fight. Brining AR QuickLook support to WebView would perhaps be the single best solution, as that would fix it for all. 

WWDC is just a few weeks away. Maybe Apple will reveal a new direction with WebXR and / or `<model>` element which might make all this a moot point.

**Reply by nabilfreeman** [reply]
Simple reason why social media apps won't use `SFSafariViewController`. It is sandboxed.

No script injection (used to track what you're doing) and content blockers (can't be switched off or bypassed inside `SFSafariViewController`.

Our only hope is for Apple to force it, which I'm sure they will in a future OS version, but resistance will be extreme.

**Reply by milesgreen** [reply]
Sadly, this. All the big social apps are using WKWebView for in-app browsers, injecting code that enables them to spy on anything the user does on visited sites. Massive security risk. Also a degraded user experience. Notable exception is Twitter iOS, that uses SFSafariViewController. 

Getting off topic here, but there's growing awareness of the significance of the problem:
https://www.forbes.com/sites/richardnieva/2022/08/18/tiktok-in-app-browser-research/
https://twitter.com/KrauseFx/status/1560370732705742848

Seeing as it's possible for a website to detect if it Is being loaded inside a WKWebView, I wish more sites would simply block in-app browsers. But that would risk a massive drop off in traffic, as users might not go the extra step to 'Open in Safari', so I can understand why they might not.

**Comment by milesgreen** [comment]
There is a bug logged on webkit requesting AR QuickLook support to be brought to `WKWebView`. This is what's needed to fix social in-app browsers (if they won't change their ways and shift to `SFSafariViewController`)

Pile on and +1 to show how important it is!

https://bugs.webkit.org/show_bug.cgi?id=240117

**Reply by rawwerks** [reply]
Does anyone know if the iOS YouTube app also uses WebKit? (In other words, will model-viewer sites load correctly if clicked inside the YT app on iOS?) 

I'm desperately in search of a channel to share my model-viewer-powered AR sites without forcing users into any of the workarounds we've outlined above...(& ideally one that isn't Twitter ;)

**Reply by elalish** [reply]
WebKit is the only browser engine allowed on iOS, so yes. Not sure if they use WKWebView; might be best to just try and see.

**Reply by rawwerks** [reply]
so from a very quick test, it appears that the current YouTube iOS app gives users a prompt for which browser to use for opening links. (yay). 
the good news is that if the user picks safari, model-viewer works.
the bad news is that if the user picks chrome, it does not. presumably because of this https://bugs.chromium.org/p/chromium/issues/detail?id=1308875

now i will just have to train everyone watching the video not to use chrome....

**Reply by elalish** [reply]
Nice; Also we've had reports that Chrome 102 on iOS is now working; if you can confirm, that would be helpful.

**Comment by rawwerks** [comment]
is anyone aware of any progress/updates re: AR for the in-app browsers? i haven't seen any indication that webkit is going to change anything.

@milesgreen?

**Reply by milesgreen** [reply]
No progress I've seen as of yet. But @grorg has indicated that Apple might someday possibly bring ARQL to WKWebView. I hope so.

**Reply by rawwerks** [reply]
thanks for the update!

i will get back to praying for this beautiful day to come. hopefully Siri is listening ;)

**Reply by rawwerks** [reply]
@Grorg - is there any hope of this happening in 2023?

**Reply by rawwerks** [reply]
@grorg - any hope of this happening in 2024? https://bugs.webkit.org/show_bug.cgi?id=240117

**Comment by zelghrabi-edu** [comment]
Any news on this, guys @milesgreen @rawwerks @elalish ?

**Reply by rawwerks** [reply]
> Any news on this, guys @milesgreen @rawwerks @elalish ?

Not as far as I can tell. WebKit team doesn't care. None of the socials other than Twitter are running SafariViewController. It's a major bummer, I've given up.

**Reply by milesgreen** [reply]
> Any news on this, guys @milesgreen @rawwerks @elalish ?

Nothing so far. Keeping my fingers crossed for WWDC on Jun 5. Hopefully there might be some improvements. Or some altogether new developments.

**Comment by marcusraty** [comment]
Came here because I too am running into this issue in April 2024. It’s quite sad to see there is no resolution here. 

Anyone tried anything that worked to get Quick Look on iOS working using in app browsers?

**Reply by rawwerks** [reply]
SFSafariViewController is the only in-app browser that I know works. (This is what X/Twitter uses.)

@milesgreen is the world's foremost expert on this problem...he may have some additional ideas.

**Reply by milesgreen** [reply]
Heh. Thanks. yeah, I've spent some time looking at this over the years...

So, the short answer is no, there's no real progress on this front. 

Best option right now is to check if the page is running inside WKWebView and if it is, pop up a message and direct the user to open in a proper browser. 

But, you need to make sure you don't block the third-party browsers that _are_ built on top of WKWebView but _have_ added support for ARQuickLook. EG. Chrome. 

So you might want to do something like:

```
if(!modelViewer.canActivateAR && window.webkit && window.webkit.messageHandlers) {
   // inside an incompatible WKWebView
}
```

If you detect an incompatible view you can:

- Show a message. Instruct them to open page in a proper browser.
- Show a 'Copy link' button and use copy to clipboard (which does work) to give them an easy way to transfer over. 
- Use a WebShare API which, I think still works in WKWebView.
